File: 20220228-hackerrank-two-two/main.py
#!/usr/bin/env python3

# https://www.hackerrank.com/challenges/two-two/problem


class Node:
    def __init__(self):
        self.is_end : bool = False
        self.next : dict[int, Node] = {}
graph = Node()


def twoTwo(a):
    count = 0
    traverses = []
    for ch in reversed(str(a)):
        i = int(ch)
        new_traverses = []
        for traverse in traverses + [graph]:
            if i not in traverse.next:
                continue
            if traverse.next[i].is_end:
                count += 1
            new_traverses.append(traverse.next[i])
        traverses = new_traverses
    return count


def main():
    # Populate "graph"
    for i in range(801):
        num = 2 ** i
        traverse = graph
        for ch in reversed(str(num)):
            i = int(ch)
            if i not in traverse.next:
                traverse.next[i] = Node()
            traverse = traverse.next[i]
        traverse.is_end = True

    T = int(input())
    for _ in range(T):
        a = input()
        print(twoTwo(a))


if __name__ == '__main__':
    main()

File: 20220228-hackerrank-two-two/main.cpp
// https://www.hackerrank.com/challenges/two-two/problem

#include <algorithm>
#include <iostream>
#include <map>
#include <string>
#include <vector>


struct Node {
    bool is_end;
    std::map<int, Node*> next;
};
Node* graph = new Node;


int twoTwo(std::string a) {
    int count = 0;
    std::vector<Node*> traverses;
    std::reverse(a.begin(), a.end());
    for (char ch : a) {
        int i = ch - '0';
        std::vector<Node*> new_traverses;
        traverses.push_back(graph);
        for (Node* traverse : traverses) {
            bool is_exist = traverse->next.find(i) != traverse->next.end();
            if (! is_exist) {
                continue;
            }
            if (traverse->next[i]->is_end) {
                count++;
            }
            new_traverses.push_back(traverse->next[i]);
        }
        traverses = new_traverses;
    }
    return count;
}


void multiply(int x, std::vector<int>& num) {
    // Multiply num * x and store the result on num[]. Note num[] stores the numbers in reversed order.
    // Ref: https://www.geeksforgeeks.org/factorial-large-number/
    int carry = 0;
    for (int& i : num) {
        int prod = i * x + carry;
        i = prod % 10;
        carry = prod / 10;
    }
    while (carry) {
        num.push_back(carry % 10);
        carry /= 10;
    }
}


int main() {
    // Populate "graph"
    std::vector<int> num = {1};
    for (int x = 0; x <= 800; x++) {
        Node* traverse = graph;
        // num[] is already in reversed order
        for (int i : num) {
            bool is_exist = traverse->next.find(i) != traverse->next.end();
            if (! is_exist) {
                traverse->next[i] = new Node;
            }
            traverse = traverse->next[i];
        }
        traverse->is_end = true;
        multiply(2, num);
    }

    int T;
    std::cin >> T;
    while (T--) {
        std::string a;
        std::cin >> a;
        std::cout << twoTwo(a) << std::endl;
    }
    return 0;
}
